### PR TITLE
Port billing and usage updates from demo branch.

### PR DESCRIPTION
**Pull request by rtamalin**
Ported demo branch changes:
  * Refactor billing functions for new config

    Rename dimensions to billed_dimensions
    Add empty metering based on reporting time
    Add bill utils module and move all relevant functions.
    Add test billing interval

  * Fix comments.

  * Add warning log for debugging local csp meter call

  * Use last bill time to calculate next bill time

Add unit test support for updated csp_cache module and the newly added
bill_utils module.

Renamed config_good.yaml to config_good_average.yaml, as it specifes
average as the usage_aggregation method.

Added a new config_good_maximum.yaml that specifies to use a maximum
usage_aggregation method.

Added a new config_test_mixed.yaml which has two metrics, one with
average as the usage_aggregation method, and the other with maximum
as the usage_aggregation method.

Updated conftest.py to reflect rename of default config file, and to
also return a Config object.

Also moved the registration of in-memory cache and csp_config hook
implementations into the conftest.py's cba_pm fixture.

Co-authored-by: smarlowucf <sean.marlow@suse.com>
Co-authored-by: rtamalin <fergal.mccarthy@suse.com>

### CODE REVIEW

**Comment by rtamalin** [comment]
With these changes I'm up to ~87.5% coverage:

```shell
---------- coverage: platform linux, python 3.10.6-final-0 -----------
Name                                       Stmts   Miss  Cover
--------------------------------------------------------------
csp_billing_adapter/__init__.py                6      0   100%
csp_billing_adapter/adapter.py                59     32    46%
csp_billing_adapter/bill_utils.py             61      4    93%
csp_billing_adapter/config.py                 28      0   100%
csp_billing_adapter/csp_cache.py              20      0   100%
csp_billing_adapter/csp_config.py              7      0   100%
csp_billing_adapter/csp_hookspecs.py          11      0   100%
csp_billing_adapter/hookspecs.py              10      1    90%
csp_billing_adapter/local_csp.py              19      0   100%
csp_billing_adapter/memory_cache.py           19      0   100%
csp_billing_adapter/memory_csp_config.py      19      0   100%
csp_billing_adapter/product_api.py             9      0   100%
csp_billing_adapter/storage_hookspecs.py      15      0   100%
csp_billing_adapter/utils.py                  30      2    93%
--------------------------------------------------------------
TOTAL                                        313     39    88%

FAIL Required test coverage of 90.0% not reached. Total coverage: 87.54%
```

File: tests/unit/test_utils.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def test_bad_get_date_delta_type_error():
     """Text exception raised for utils.get_date_delta"""
     with raises(TypeError, match="foo"):
         utils.get_date_delta(datetime.datetime(2023, 4, 20, 13, 8, 30),
-        'foo')
+                             'foo')
```
Comment (brett060102) [inline tests/unit/test_utils.py:75]: repeat Sean's nit about new lines.

**Review by brett060102 — APPROVED** [review]
looks good. minor nit.

File: csp_billing_adapter/adapter.py [hunk 2]
```diff
@@ -25,23 +25,18 @@
...
 from csp_billing_adapter import csp_hookspecs
 from csp_billing_adapter import storage_hookspecs
 from csp_billing_adapter import (
     local_csp,
-    product_api,
-    memory_cache,
-    memory_csp_config
 )
 
 default_config_path = '/etc/csp_billing_adapter/config.yaml'
```
Comment (KeithMnemonic) [inline csp_billing_adapter/adapter.py:42]: should this be all caps ? I probably did this lower case and did not consider it should be caps

**Review by KeithMnemonic — APPROVED** [review]
LGTM, one small comment

File: tests/unit/test_utils.py (hunk 1 above)
Comment (rtamalin) [reply tests/unit/test_utils.py:75]: Not sure I understand the comment here?

This is a minimal change addressing the existing flake8 complaint that I saw when I ran it against the tests area.

Comment (rtamalin) [reply tests/unit/test_utils.py]: I changed it to the more fully expanded way of writing this as part of my cleanup & updates to the commit message.

File: csp_billing_adapter/adapter.py (hunk 2 above)
Comment (rtamalin) [reply csp_billing_adapter/adapter.py:42]: Yeah since this should be considered a constant... I can fix it here if you want?

Comment (KeithMnemonic) [reply csp_billing_adapter/adapter.py:42]: please fix , thank you

Comment (rtamalin) [reply csp_billing_adapter/adapter.py]: Done

**Comment by rtamalin** [comment]
With a couple of last minute tweaks to test_bill_utils.py and test_utils.py, the coverage is up to 88.82%.

```shell
% pytest --cov=csp_billing_adapter --cov-report term-missing
========================================================= test session starts ==========================================================
platform linux -- Python 3.10.6, pytest-7.3.1, pluggy-1.0.0
rootdir: /home/fergal/work/public-cloud-team/csp-adapter/github/csp-billing-adapter
configfile: setup.cfg
testpaths: tests
plugins: cov-4.0.0
collected 40 items                                                                                                                     

tests/unit/test_bill_utils.py ........                                                                                           [ 20%]
tests/unit/test_config.py ...                                                                                                    [ 27%]
tests/unit/test_csp_cache.py ....                                                                                                [ 37%]
tests/unit/test_csp_config.py .                                                                                                  [ 40%]
tests/unit/test_local_csp.py ....                                                                                                [ 50%]
tests/unit/test_memory_cache.py ....                                                                                             [ 60%]
tests/unit/test_memory_csp_config.py ....                                                                                        [ 70%]
tests/unit/test_product_api.py .                                                                                                 [ 72%]
tests/unit/test_utils.py ...........                                                                                             [100%]

---------- coverage: platform linux, python 3.10.6-final-0 -----------
Name                                       Stmts   Miss  Cover   Missing
------------------------------------------------------------------------
csp_billing_adapter/__init__.py                6      0   100%
csp_billing_adapter/adapter.py                59     32    46%   58-102, 106
csp_billing_adapter/bill_utils.py             61      2    97%   68, 95
csp_billing_adapter/config.py                 28      0   100%
csp_billing_adapter/csp_cache.py              20      0   100%
csp_billing_adapter/csp_config.py              7      0   100%
csp_billing_adapter/csp_hookspecs.py          11      0   100%
csp_billing_adapter/hookspecs.py              10      1    90%   45
csp_billing_adapter/local_csp.py              19      0   100%
csp_billing_adapter/memory_cache.py           19      0   100%
csp_billing_adapter/memory_csp_config.py      19      0   100%
csp_billing_adapter/product_api.py             9      0   100%
csp_billing_adapter/storage_hookspecs.py      15      0   100%
csp_billing_adapter/utils.py                  30      0   100%
------------------------------------------------------------------------
TOTAL                                        313     35    89%

FAIL Required test coverage of 90.0% not reached. Total coverage: 88.82%

========================================================== 40 passed in 0.40s ==========================================================
```

**Review by smarlowucf — APPROVED** [review]
LGTM

**Pull request merged**